Tutorial: How to set up a public beta program using early access management (#6403)
PostHog/posthog.com: contents/tutorials/public-beta-program.md

---
title: How to set up a public beta program using early access management
date: 2023-07-28
author: ["ian-vanagas"]
showTitle: true
sidebar: Docs
featuredImage: ../images/tutorials/banners/tutorial-10.png
tags: ['feature flags']
---

> Testing your new features helps make them successful. Public betas are a way to get new features into the hands of users and get valuable feedback and analytics. In this tutorial, we show you how to add a basic public beta program to your (Next.js) app with PostHog’s early access management feature.

Reviewer: Perhaps worth mentioning that while this is Next.js, they can follow something similar for other frameworks?

```suggestion
Public betas are a way to get new features into the hands of users and get valuable feedback and analytics. In this tutorial, we show you how to add a basic public beta program to your app with PostHog’s early access management feature.

Our example in this tutorial focuses on a Next.js app. However, similar principles apply if you're using a different language or framework.
```

> Already have an app and PostHog set up? [Skip creating the beta feature](#creating-our-beta-feature-in-posthog).

## Creating a Next.js app and adding PostHog

> First, we need an app and we use Next.js to create one. After installing Node, run the command below, select **No** for TypeScript, **Yes** for `use app router`, and the defaults for every other option.

Reviewer: ```suggestion
First, create a Next.js app. After installing Node, run the command below, select **No** for TypeScript, **Yes** for `use app router`, and the defaults for every other option.
```

## Creating a beta opt-in page

For full control over our beta opt-in page, we can create a custom page instead of using the modal.

To do this, first, create a new folder named `beta`. In this `beta` folder, create a file named `page.js`. In this file, we will:

1. Check for available betas and active feature flags.
2. Create two lists: currently active betas and currently inactive betas.
3. Display each list on the page with buttons to opt in or out of the beta.
4. Toggle their early access enrollment when they click those buttons.

First, set up a basic page with PostHog, our active flags, states for active and inactive betas, and spots to display both of them.

```js
// app/beta/page.js
'use client'
import { useActiveFeatureFlags, usePostHog } from 'posthog-js/react'
import { useState, useEffect } from 'react'

export default function Beta() {

  const posthog = usePostHog()
  const activeFlags = useActiveFeatureFlags()

  const [activeBetas, setActiveBetas] = useState([])
  const [inactiveBetas, setInactiveBetas] = useState([])
  
  return (
    <div>
      <h1>Available betas</h1>
      <h3>Inactive</h3>
      <h3>Active</h3>
    </div>
  )
}
```

> Next, we set up a `useEffect` for our key logic. It filters the early access features (from `getEarlyAccessFeatures`) into active and inactive betas depending on if the flag key exists and then sets those two filtered lists in their states. We also just set all betas to inactive if they don’t have any active flags.

Reviewer: ```suggestion
Next, we set up a `useEffect` hook for our key logic. It filters the early access features (from `getEarlyAccessFeatures`) into active and inactive betas depending on if the flag key exists and then sets those two filtered lists in their states. We also just set all betas to inactive if they don’t have any active flags.
```
